### PR TITLE
[TFF-226] Refactor model data

### PR DESCRIPTION
**Pull request by thofmann**
### Summary

**>>> Test and merge with the sister PR! TheFourFifths/consus-client/pull/114 <<<**

This refactors the data model for models which have been checked out. Rather than storing them as `[model1, model1, model1, model2]`, they'll be stored as `[{address: model1.address, quantity: 3}, {address: model2.address, quantity: 1}]`. This makes them much easier to work with.

### Checklist

- [x] Have you added unit and functional tests as appropriate?
- [x] Did you update/add documentation for new methods or changed functionality?
- [x] Have you merged the target branch down into this one?

### How to Review

1. Review the code for sanity.
2. Test the client and server, both on the `refactor-model-TFF-226` branch, to manually make sure that everything still works (particularly model check-in and checkout!).
3. Make sure I didn't miss any needed documentation changes.

### Links

- [TFF-226](https://msoese.atlassian.net/browse/TFF-226)
- TheFourFifths/consus-client/pull/114

### CODE REVIEW

File: src/api/checkout.js [hunk 1]
```diff
@@ -7,19 +7,16 @@ app.post('/', (req, res) => {
...
-    if (typeof req.body.equipmentAddresses !== 'object'
-        || typeof req.body.equipmentAddresses[0] !== 'string') {
-        return res.failureJson('An array of item addresses is required.');
+    if (typeof req.body.equipment !== 'object') {
+        return res.failureJson('An array of equipment is required.');
     }
     addAction('NEW_CHECKOUT', {
         studentId: req.body.studentId,
-        equipmentAddresses: req.body.equipmentAddresses,
+        equipment: req.body.equipment,
```
Comment (mattmahn) [inline src/api/checkout.js]: NACK: Update docs/Actions.md

File: src/api/checkout.js [hunk 2]
```diff
@@ -28,16 +25,15 @@ app.post('/longterm', (req, res) => {
...
+    if (typeof req.body.equipment !== 'object') {
+        return res.failureJson('An array of equipment is required.');
     }
     if (typeof req.body.professor !== 'string') {
         return res.failureJson('A professor is required.');
     }
     addAction('NEW_LONGTERM_CHECKOUT', {
         studentId: req.body.studentId,
-        equipmentAddresses: req.body.equipmentAddresses,
+        equipment: req.body.equipment,
```
Comment (mattmahn) [inline src/api/checkout.js]: NACK: update docs/Actions.md

File: src/lib/clock.js [hunk 3]
```diff
@@ -24,3 +24,24 @@ setTimeout(() => {
 export function isBeforeNow(timestamp){
     return timestamp < Date.now() / 1000;
 }
+
+export function dueDateToTimestamp(dueDate) {
```
Comment (mattmahn) [inline src/lib/clock.js]: :+1: :100: :1st_place_medal:

File: src/store/student-store.js [hunk 4]
```diff
@@ -81,18 +80,31 @@ function removeItemFromAllStudents(itemAddress) {
...
+    equipment.forEach(equip => {
+        let address = equip.address;
+        let type = readAddress(address).type;
+        if (type === 'item') {
             student.items.push(ItemStore.getItemByAddress(address));
-        }
-        else if (result.type == 'model') {
-            student.models.push(ModelStore.getModelByAddress(address));
+        } else if (type === 'model') {
+            let index = student.models.findIndex(model => model.address === address);
```
Comment (mattmahn) [inline src/store/student-store.js]: Nit: why not just `let model = student.models.find(model => model.address === address);`?

File: src/store/student-store.js [hunk 5]
```diff
@@ -115,12 +127,12 @@ store.registerHandler('NEW_STUDENT', data => {
 
 store.registerHandler('NEW_CHECKOUT', data => {
     store.waitFor(CheckoutStore);
-    addItemsToStudent(data.studentId, data.equipmentAddresses);
+    addItemsToStudent(data.studentId, data.equipment, data.timestamp);
```
Comment (mattmahn) [inline src/store/student-store.js]: NACK: docs/Actions.md: `data.timestamp`

File: test/unit/store/student-store.js [hunk 6]
```diff
@@ -102,7 +102,17 @@ describe('StudentStore', () => {
...
-            equipmentAddresses: [items[0].address, items[2].address, models[0].address]
+            equipment: [
+                {
+                    address: items[0].address
+                },
+                {
+                    address: items[2].address
+                },
+                {
+                    address: models[0].address
```
Comment (mattmahn) [inline test/unit/store/student-store.js]: Are you missing `quantity`?

File: src/store/student-store.js (hunk 4 above)
Comment (thofmann) [inline src/store/student-store.js]: Good question; will change.

File: src/api/checkout.js (hunk 1 above)
Comment (thofmann) [inline src/api/checkout.js]: Updated

File: src/api/checkout.js (hunk 2 above)
Comment (thofmann) [inline src/api/checkout.js]: Updated

File: src/store/student-store.js (hunk 5 above)
Comment (thofmann) [inline src/store/student-store.js]: `data.timestamp` is automatically added to all actions. I don't think we need to document this.

File: test/unit/store/student-store.js (hunk 6 above)
Comment (thofmann) [inline test/unit/store/student-store.js]: I am missing `quantity`. Good catch.

Fixed.

File: docs/Actions.md [hunk 7]
```diff
@@ -9,6 +9,7 @@ This document describes the Flux actions used in the Consus client.
 - [Actions](#actions)
     - [Table of contents](#table-of-contents)
     - [CHECKIN](#checkin)
+    - [CHECKIN_MODELS](#checkin_models)
```
Comment (tondef001) [inline docs/Actions.md]: Am I crazy?  Didn't I already add this?

File: src/store/checkout-store.js [hunk 8]
```diff
@@ -32,17 +32,18 @@ class CheckoutStore extends Store {
...
 const store = new CheckoutStore();
 
-function verifyEquipmentAvailability(equipmentAddresses){
-    equipmentAddresses.forEach(address => {
+function verifyEquipmentAvailability(equipment){
+    equipment.forEach(equip => {
+        let address = equip.address;
         let result = readAddress(address);
-        if(result.type == 'item'){
+        if (result.type == 'item'){
```
Comment (gearedmonkey) [inline src/store/checkout-store.js]: Strict equal?

File: src/store/checkout-store.js [hunk 9]
```diff
@@ -32,17 +32,18 @@ class CheckoutStore extends Store {
...
             if (ItemStore.getItemByAddress(address).status !== 'AVAILABLE') {
                 throw new Error('An item in the cart is not available for checkout.');
             }
         }
-        if(result.type == 'model'){
+        if (result.type == 'model'){
```
Comment (gearedmonkey) [inline src/store/checkout-store.js]: Strict equal?

File: src/store/checkout-store.js (hunk 8 above)
Comment (thofmann) [inline src/store/checkout-store.js]: Wasn't me. Will fix.

File: src/store/checkout-store.js (hunk 9 above)
Comment (thofmann) [inline src/store/checkout-store.js]: Wasn't me. Will fix.

**Comment by thofmann** [comment]
>A normal longterm checkout does not work anymore. The date is incorrectly set. Verified that it is functional on DEV

Good catch @gearedmonkey. Working on this.

**Pull request merged**